pkg/controller/directvolumemigration/rsync.go

```diff
@@ -1660,3 +1690,683 @@ func (t *Task) deleteProgressReportingCRs(client k8sclient.Client) error {
+	}
+	r.ops = append(r.ops, s)
+}
+
+// AllCompleted checks whether all of the Rsync attempts are in a terminal state
+// If true, reconcile can move to next phase
+func (r *rsyncClientOperationStatusList) AllCompleted() bool {
+	for _, attempt := range r.ops {
+		if attempt.pending || attempt.running || attempt.HasErrors() {
+			return false
+		}
+	}
+	return true
+}
+
+// AnyErrored checks whether any of the operation is resulting in an error
+func (r *rsyncClientOperationStatusList) AnyErrored() bool {
+	for _, attempt := range r.ops {
+		if attempt.HasErrors() {
+			return true
+		}
+	}
+	return false
+}
+
+// Failed returns number of failed operations
+func (r *rsyncClientOperationStatusList) Failed() int {
+	i := 0
+	for _, attempt := range r.ops {
+		if attempt.failed {
+			i += 1
+		}
+	}
+	return i
+}
+
+// Succeeded returns number of failed operations
+func (r *rsyncClientOperationStatusList) Succeeded() int {
+	i := 0
+	for _, attempt := range r.ops {
+		if attempt.succeeded {
+			i += 1
+		}
+	}
+	return i
+}
+
+// Pending returns number of pending operations
+func (r *rsyncClientOperationStatusList) Pending() int {
+	i := 0
+	for _, attempt := range r.ops {
+		if attempt.pending {
+			i += 1
+		}
+	}
+	return i
+}
+
+// Running returns number of running operations
+func (r *rsyncClientOperationStatusList) Running() int {
+	i := 0
+	for _, attempt := range r.ops {
+		if attempt.pending {
+			i += 1
+		}
+	}
+	return i
+}
+
+// ensureRsyncOperations orchestrates all attempts of Rsync, updates owner status with observed state of all ongoing Rsync operations
+// returns structured status of all operations, the return value should be used to make decisions about whether to retry in next reconcile
+func (t *Task) ensureRsyncOperations(client compat.Client, podRequirements []rsyncClientPodRequirements) (rsyncClientOperationStatusList, []error) {
+	statusList := rsyncClientOperationStatusList{}
+	// rateLimiter defines maximum concurrent operations that can be reconciled in one go
+	operationsRateLimiter, garbageCollectionRateLimiter := make(chan bool, DefaultRsyncOperationConcurrency), make(chan bool, 2)
+	// outputChan streamlines output of concurrent reconcile operations
+	operationOutputChan := make(chan rsyncClientOperationStatus, DefaultRsyncOperationConcurrency)
+	garbageCollectionOutputChan := make(chan []error, 2)
+	finishOperationChan, finishGarbageCollectionChan := make(chan bool), make(chan bool)
+	garbageCollectionErrors := []error{}
+	waitGroup := &sync.WaitGroup{}
+	mutex := &sync.Mutex{}
+	for i := range podRequirements {
+		req := &podRequirements[i]
+		lastObservedOperationStatus := t.Owner.Status.GetRsyncOperationStatusForPVC(&corev1.ObjectReference{
+			Name:      req.pvInfo.name,
+			Namespace: req.namespace,
+		})
+		// from this point onwards, do not mutate the original reference, create a copy and use it
+		threadSafeOperationStatus := *lastObservedOperationStatus.DeepCopy()
```

Comment:
@djwhatle I investigated the wrapper around api fields and usage for thread-safe access methods. I ran into some problems with the encoder not being able to reflect into fields. Therefore, I decided to make it outside of API. The above deep copy here makes sure that every goroutine gets their own copy of Observed Status. I changed the functions to use copy of the variable instead of taking references, which makes sure there will never be concurrent reads/writes on the data structures used as they are no longer shared.